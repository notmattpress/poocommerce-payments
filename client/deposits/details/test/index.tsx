/** @format */

/**
 * External dependencies
 */
import { render } from '@testing-library/react';
import React from 'react';

/**
 * Internal dependencies
 */
import type { CachedDeposit } from 'types/deposits';
import { DepositOverview } from '../';

const mockDeposit = {
	id: 'po_mock',
	date: '2020-01-02 17:46:02',
	type: 'deposit',
	amount: 2000,
	status: 'paid',
	bankAccount: 'MOCK BANK •••• 1234 (USD)',
	automatic: true,
	fee: 30,
	fee_percentage: 1.5,
	currency: 'USD',
} as CachedDeposit;

const mockWithdrawal = {
	id: 'po_mock',
	date: '2020-01-02 17:46:02',
	type: 'withdrawal',
	amount: -2000,
	status: 'paid',
	bankAccount: 'MOCK BANK •••• 1234 (USD)',
	automatic: true,
	fee: 30,
	fee_percentage: 1.5,
	currency: 'USD',
} as CachedDeposit;

declare const global: {
	wcpaySettings: {
		zeroDecimalCurrencies: string[];
		currencyData: Record< string, any >;
		connect: {
			country: string;
		};
		dateFormat: string;
	};
	wcSettings: { countries: Record< string, string > };
};

describe( 'Deposit overview', () => {
	beforeEach( () => {
		jest.clearAllMocks();
		global.wcpaySettings = {
			zeroDecimalCurrencies: [],
			connect: {
				country: 'US',
			},
			currencyData: {
				US: {
					code: 'USD',
					symbol: '$',
					symbolPosition: 'left',
					thousandSeparator: ',',
					decimalSeparator: '.',
					precision: 2,
				},
			},
			dateFormat: 'M j, Y',
		};
	} );

	test( 'renders automatic payout correctly', () => {
		const { container: overview, getByText } = render(
			<DepositOverview deposit={ mockDeposit } />
		);
		getByText( /Payout date:/ );
		getByText( 'Completed (paid)' );
		expect( overview ).toMatchSnapshot();
	} );

	test( 'renders automatic withdrawal correctly', () => {
		const { container: overview, getByText } = render(
			<DepositOverview deposit={ mockWithdrawal } />
		);
		getByText( /Withdrawal date:/ );
		getByText( 'Completed (deducted)' );
		expect( overview ).toMatchSnapshot();
	} );

	test( 'renders instant deposit correctly', () => {
		const { container: overview } = render(
			<DepositOverview deposit={ { ...mockDeposit, automatic: false } } />
		);
		expect( overview ).toMatchSnapshot();
	} );

	// test when deposit data could not be found, it renders a notice
	test( 'renders notice when deposit data is not found', () => {
		const { container: overview } = render(
			<DepositOverview deposit={ undefined } />
		);
		expect( overview ).toMatchSnapshot();
	} );

	test( 'renders failure reason when deposit has a known failure code', () => {
		const failedDeposit = {
			...mockDeposit,
			status: 'failed',
			failure_code: 'insufficient_funds',
		} as CachedDeposit;

		const { getByText } = render(
			<DepositOverview deposit={ failedDeposit } />
		);

		expect( getByText( 'Failure reason:' ) ).toBeInTheDocument();
		expect(
			getByText(
				'Your account has insufficient funds to cover your negative balance.'
			)
		).toBeInTheDocument();
	} );

	test( 'renders failure_message when failure_code is new and not included in our mapping', () => {
		// @ts-expect-error Testing invalid failure code scenario
		const failedDeposit = {
			...mockDeposit,
			status: 'failed',
			failure_code: 'unknown_failure_code',
			failure_message:
				'Failure error message originally captured from the Stripe Payout object',
		} as CachedDeposit;

		const { getByText } = render(
			<DepositOverview deposit={ failedDeposit } />
		);

		expect( getByText( 'Failure reason:' ) ).toBeInTheDocument();
		expect(
			getByText(
				'Failure error message originally captured from the Stripe Payout object'
			)
		).toBeInTheDocument();
	} );

	test( 'renders failure_message when no failure_code exists', () => {
		const failedDeposit = {
			...mockDeposit,
			status: 'failed',
			failure_message:
				'Failure error message originally captured from the Stripe Payout object',
		} as CachedDeposit;

		const { getByText } = render(
			<DepositOverview deposit={ failedDeposit } />
		);

		expect( getByText( 'Failure reason:' ) ).toBeInTheDocument();
		expect(
			getByText(
				'Failure error message originally captured from the Stripe Payout object'
			)
		).toBeInTheDocument();
	} );

	test( 'renders Unknown when no failure_code nor failure_message exist - edge case', () => {
		const failedDeposit = {
			...mockDeposit,
			status: 'failed',
		} as CachedDeposit;

		const { getByText } = render(
			<DepositOverview deposit={ failedDeposit } />
		);

		expect( getByText( 'Failure reason:' ) ).toBeInTheDocument();
		expect( getByText( 'Unknown' ) ).toBeInTheDocument();
	} );
} );
