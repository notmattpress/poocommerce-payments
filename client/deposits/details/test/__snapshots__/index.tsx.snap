// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deposit overview renders automatic payout correctly 1`] = `
<div>
  <div
    class="wcpay-deposit-overview"
  >
    <div
      class="components-surface components-card wcpay-deposit-automatic css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <ul>
          <li
            class="woocommerce-summary__item-container"
          >
            <div
              class="woocommerce-summary__item"
            >
              <div
                class="woocommerce-summary__item-label"
              >
                Payout date: Jan 2, 2020
              </div>
              <div
                class="woocommerce-summary__item-data"
              >
                <div
                  class="woocommerce-summary__item-value"
                >
                  <div
                    class="woocommerce-order-status"
                  >
                    <span
                      class="woocommerce-order-status__indicator is-paid"
                    />
                    Completed (paid)
                  </div>
                </div>
              </div>
            </div>
          </li>
          <li
            class="wcpay-deposit-amount"
          >
            $20.00
          </li>
        </ul>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
    <div
      class="components-surface components-card css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-flex components-card__header components-card-header css-1g5oj2q-View-Flex-sx-Base-sx-Items-ItemsRow-Header-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardHeader"
        >
          <span
            class="components-truncate components-text css-1sg6qaf-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Payout details
          </span>
        </div>
        <div
          class="components-card__body components-card-body css-1nwhnu3-View-Body-borderRadius-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            class="woopayments-payout-details-header"
          >
            <div
              class="woopayments-payout-details-header__item"
            >
              <h2>
                Bank account
              </h2>
              <div
                class="woopayments-payout-details-header__value"
              >
                MOCK BANK •••• 1234 (USD)
              </div>
            </div>
            <div
              class="woopayments-payout-details-header__item"
            >
              <h2>
                Bank reference ID
              </h2>
              <div
                class="woopayments-payout-details-header__value"
              >
                <div
                  class="woopayments-payout-details-header__value"
                >
                  Not available
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
  </div>
</div>
`;

exports[`Deposit overview renders automatic withdrawal correctly 1`] = `
<div>
  <div
    class="wcpay-deposit-overview"
  >
    <div
      class="components-surface components-card wcpay-deposit-automatic css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <ul>
          <li
            class="woocommerce-summary__item-container"
          >
            <div
              class="woocommerce-summary__item"
            >
              <div
                class="woocommerce-summary__item-label"
              >
                Withdrawal date: Jan 2, 2020
              </div>
              <div
                class="woocommerce-summary__item-data"
              >
                <div
                  class="woocommerce-summary__item-value"
                >
                  <div
                    class="woocommerce-order-status"
                  >
                    <span
                      class="woocommerce-order-status__indicator is-paid"
                    />
                    Completed (deducted)
                  </div>
                </div>
              </div>
            </div>
          </li>
          <li
            class="wcpay-deposit-amount"
          >
            -$20.00
          </li>
        </ul>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
    <div
      class="components-surface components-card css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-flex components-card__header components-card-header css-1g5oj2q-View-Flex-sx-Base-sx-Items-ItemsRow-Header-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardHeader"
        >
          <span
            class="components-truncate components-text css-1sg6qaf-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Withdrawal details
          </span>
        </div>
        <div
          class="components-card__body components-card-body css-1nwhnu3-View-Body-borderRadius-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            class="woopayments-payout-details-header"
          >
            <div
              class="woopayments-payout-details-header__item"
            >
              <h2>
                Bank account
              </h2>
              <div
                class="woopayments-payout-details-header__value"
              >
                MOCK BANK •••• 1234 (USD)
              </div>
            </div>
            <div
              class="woopayments-payout-details-header__item"
            >
              <h2>
                Bank reference ID
              </h2>
              <div
                class="woopayments-payout-details-header__value"
              >
                <div
                  class="woopayments-payout-details-header__value"
                >
                  Not available
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
  </div>
</div>
`;

exports[`Deposit overview renders instant deposit correctly 1`] = `
<div>
  <div
    class="wcpay-deposit-overview"
  >
    <div
      aria-describedby="woocommerce-summary-helptext-1"
      aria-label="Payout overview"
      aria-orientation="horizontal"
      role="menu"
    >
      <p
        class="screen-reader-text"
        id="woocommerce-summary-helptext-1"
      >
        List of data points available for filtering. Use arrow keys to cycle through the list. Click a data point for a detailed report.
      </p>
      <ul
        class="woocommerce-summary has-4-items"
      >
        <li
          class="woocommerce-summary__item-container"
        >
          <div
            class="woocommerce-summary__item"
          >
            <div
              class="woocommerce-summary__item-label"
            >
              Instant payout date: Jan 2, 2020
            </div>
            <div
              class="woocommerce-summary__item-data"
            >
              <div
                class="woocommerce-summary__item-value"
              >
                <div
                  class="woocommerce-order-status"
                >
                  <span
                    class="woocommerce-order-status__indicator is-paid"
                  />
                  Completed (paid)
                </div>
              </div>
            </div>
          </div>
        </li>
        <li
          class="woocommerce-summary__item-container"
        >
          <div
            class="woocommerce-summary__item"
          >
            <div
              class="woocommerce-summary__item-label"
            >
              Payout amount
            </div>
            <div
              class="woocommerce-summary__item-data"
            >
              <div
                class="woocommerce-summary__item-value"
              >
                $20.30
              </div>
            </div>
          </div>
        </li>
        <li
          class="woocommerce-summary__item-container"
        >
          <div
            class="woocommerce-summary__item"
          >
            <div
              class="woocommerce-summary__item-label"
            >
              1.5% service fee
            </div>
            <div
              class="woocommerce-summary__item-data"
            >
              <div
                class="woocommerce-summary__item-value wcpay-deposit-fee"
              >
                $0.30
              </div>
            </div>
          </div>
        </li>
        <li
          class="woocommerce-summary__item-container"
        >
          <div
            class="woocommerce-summary__item"
          >
            <div
              class="woocommerce-summary__item-label"
            >
              Net payout amount
            </div>
            <div
              class="woocommerce-summary__item-data"
            >
              <div
                class="woocommerce-summary__item-value wcpay-deposit-net"
              >
                $20.00
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="components-surface components-card css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-flex components-card__header components-card-header css-1g5oj2q-View-Flex-sx-Base-sx-Items-ItemsRow-Header-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardHeader"
        >
          <span
            class="components-truncate components-text css-1sg6qaf-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Payout details
          </span>
        </div>
        <div
          class="components-card__body components-card-body css-1nwhnu3-View-Body-borderRadius-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            class="woopayments-payout-details-header"
          >
            <div
              class="woopayments-payout-details-header__item"
            >
              <h2>
                Bank account
              </h2>
              <div
                class="woopayments-payout-details-header__value"
              >
                MOCK BANK •••• 1234 (USD)
              </div>
            </div>
            <div
              class="woopayments-payout-details-header__item"
            >
              <h2>
                Bank reference ID
              </h2>
              <div
                class="woopayments-payout-details-header__value"
              >
                <div
                  class="woopayments-payout-details-header__value"
                >
                  Not available
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
  </div>
</div>
`;

exports[`Deposit overview renders notice when deposit data is not found 1`] = `
<div>
  <div
    class="wcpay-inline-notice wcpay-inline-error-notice components-notice is-error"
  >
    <div
      class="components-notice__content"
    >
      <div
        class="components-flex css-bmzg3m-View-Flex-sx-Base-sx-Items-ItemsRow em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Flex"
      >
        <div
          class="components-flex-item wcpay-inline-notice__icon wcpay-inline-error-notice__icon css-mw3lhz-View-Item-sx-Base em57xhy0"
          data-wp-c16t="true"
          data-wp-component="FlexItem"
        >
          <svg
            class="gridicon gridicons-notice-outline"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M12 4c4.411 0 8 3.589 8 8s-3.589 8-8 8-8-3.589-8-8 3.589-8 8-8m0-2C6.477 2 2 6.477 2 12s4.477 10 10 10 10-4.477 10-10S17.523 2 12 2zm1 13h-2v2h2v-2zm-2-2h2l.5-6h-3l.5 6z"
              />
            </g>
          </svg>
        </div>
        <div
          class="components-flex-item wcpay-inline-notice__content wcpay-inline-error-notice__content css-mw3lhz-View-Item-sx-Base em57xhy0"
          data-wp-c16t="true"
          data-wp-component="FlexItem"
        >
          The deposit you are looking for cannot be found.
        </div>
      </div>
      <div
        class="components-notice__actions"
      />
    </div>
  </div>
</div>
`;
